sdcflows/fieldmaps.py

```diff
@@ -0,0 +1,345 @@
+        """Validate a fieldmap path."""
+        if isinstance(value, BIDSFile):
+            value = Path(value.path)
+        if isinstance(value, str):
+            value = Path(value)
+
+        if not value.is_file():
+            raise FileNotFoundError(
+                f"File path <{value}> does not exist, is a broken link, or it is not a file"
+            )
+
+        if not str(value).endswith((".nii", ".nii.gz")):
+            raise ValueError(f"File path <{value}> does not look like a NIfTI file.")
+
+        suffix = str(value).rpartition(".nii")[0].rpartition("_")[-1]
+        if suffix not in tuple(MODALITIES.keys()):
+            raise ValueError(
+                f"File path <{value}> with suffix <{suffix}> is not a valid "
+                "fieldmap sourcefile."
+            )
+
+    def __attrs_post_init__(self):
+        """Validate metadata and additional checks."""
+        self.entities = parse_file_entities(str(self.path))
+        self.suffix = self.entities.pop("suffix")
+        extension = self.entities.pop("extension").lstrip(".")
+
+        # Automatically fill metadata in when possible
+        # TODO: implement BIDS hierarchy of metadata
+        if self.find_meta:
+            sidecar = Path(str(self.path).replace(extension, "json"))
+            if sidecar.is_file():
+                _meta = self.metadata or {}
+                self.metadata = loads(sidecar.read_text())
+                self.metadata.update(_meta)
+
+        # Attempt to infer a bids_root folder
+        relative_path = relative_to_root(self.path)
+        if str(relative_path) != str(self.path):
+            self.bids_root = Path(str(self.path)[: -len(str(relative_path))])
+
+        # Check for REQUIRED metadata (depends on suffix.)
+        if self.suffix in ("bold", "dwi", "epi", "sbref"):
+            if "PhaseEncodingDirection" not in self.metadata:
+                raise MetadataError(
+                    f"Missing 'PhaseEncodingDirection' for <{self.path}>."
+                )
+            if not (
+                set(("TotalReadoutTime", "EffectiveEchoSpacing")).intersection(
+                    self.metadata.keys()
+                )
+            ):
+                raise MetadataError(
+                    f"Missing readout timing information for <{self.path}>."
+                )
+
+        if self.suffix == "fieldmap" and "Units" not in self.metadata:
+            raise MetadataError(f"Missing 'Units' for <{self.path}>.")
+
+        if self.suffix == "phasediff" and (
+            "EchoTime1" not in self.metadata or "EchoTime2" not in self.metadata
+        ):
+            raise MetadataError(
+                f"Missing 'EchoTime1' and/or 'EchoTime2' for <{self.path}>."
+            )
+
+        if self.suffix in ("phase1", "phase2") and ("EchoTime" not in self.metadata):
+            raise MetadataError(f"Missing 'EchoTime' for <{self.path}>.")
+
+
+@attr.s(slots=True)
+class FieldmapEstimation:
```

Comment:
is there a reason for this to be a class, rather than a function returning sources / FM estimation method?

Reply:
The attribute `method` keeps a state of the estimation (i.e. either unknown or a type of estimation). Also, I think a class can achieve similar things to a function, while having the responsibilities more clear and readable.

Reply:
Usually when I see a class with the only method being initialization, I think it can be stripped down to a function (`estimate_fieldmap_method()`) and achieve the same amount of clarity / readability, with the benefit of reducing some bloat. Would `method` ever be altered after this has been initialized?

Reply:
The function you suggest already exists - https://github.com/nipreps/sdcflows/blob/fef4ac623910b08d1f95d47d397d1c19dad0e45c/sdcflows/workflows/base.py#L259-L295

It is all but pretty or reliable. And yet, it doesn't implement most of the validation checkpoints in this PR or the metadata. Then the staging of the actual estimation happens on a different function - https://github.com/nipreps/sdcflows/blob/fef4ac623910b08d1f95d47d397d1c19dad0e45c/sdcflows/workflows/base.py#L22-L256 that has other validation points and decisions.

I wanted to start these objects because I was uncomfortable with the existing code. That code was also not very convenient for standalone implementations (#8) and while it seems to work acceptably for fMRIPrep (acceptable because, for instance, we currently replicate the processing of the same fieldmap for all runs in a dataset, if all are pointed by just one `IntendedFor` (i.e., one fieldmap). However, when I'm trying to use this on dMRIPrep, well it just does not work.

Objects are definitely easier to test, easier to document (sphinx does a fairly good job collecting docstrings) and more maintainable in the long run because they only have one, simple objective.

One thing I'm still internally debating is whether `FieldmapEstimation` should have a `get_workflow` or such, that directly returns the nipype workflow object built and with the inputs and metadata set (i.e., ready to connect the output to some derivatives sink workflow and correction workflows). I'm leaning towards doing it. Any feedback on this will be very very much appreciated.